File: dotnet/GHIElectronics.DUELink/Button.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static GHIElectronics.DUELink.DUELinkController;

namespace GHIElectronics.DUELink {

    public partial class DUELinkController {
      
        public class ButtonController {


            SerialInterface serialPort;
            public ButtonController(SerialInterface serialPort) => this.serialPort = serialPort;


            public bool Enable(int pin, int state) {
  
                var cmd = string.Format("btnen({0},{1})", pin, state);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;

            }

            public bool Down(int pin) {
        

                var cmd = string.Format("btndown({0})", pin);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                if (res.success) {
                    try {
                        var ready = int.Parse(res.response);
                        return ready == 1 ? true : false ;
                    }
                    catch {
                       
                    }

                }

                return false;
            }

            public bool Up(int pin) {

                var cmd = string.Format("btnup({0})", pin);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                if (res.success) {
                    try {
                        var ready = int.Parse(res.response);
                        return ready == 1 ? true : false;
                    }
                    catch {

                    }
                }

                return false;
            }

            public bool Read(int pin) {
                var cmd = string.Format("btnread({0})", pin);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                if (res.success) {
                    try {
                        var ready = int.Parse(res.response);
                        return ready == 1 ? true : false;
                    }
                    catch {

                    }
                }

                return false;
            }
        }
    }
}

File: micropython/test/duelink/button.py

class ButtonController:    

    def __init__(self, serialPort):
        self.serialPort = serialPort

    def Enable(self, pin: int, state: int) -> bool:      
        cmd = f"btnen({pin}, {int(state)})"

        self.serialPort.WriteCommand(cmd)
        r,s = self.serialPort.ReadResponse()

        return r
            
    def Up(self, pin):
        cmd = f"btnup({pin})"

        self.serialPort.WriteCommand(cmd)
        r,s = self.serialPort.ReadResponse()

        if r:
            try:
                return int(s) == 1
            except:
                pass

        return False   
    
    def Down(self, pin):
        cmd = f"btndown({pin})"

        self.serialPort.WriteCommand(cmd)
        r,s = self.serialPort.ReadResponse()

        if r:
            try:
                return int(s) == 1
            except:
                pass

        return False
        
    def Read(self, pin):
        cmd = f"btnread({pin})"

        self.serialPort.WriteCommand(cmd)
        r,s = self.serialPort.ReadResponse()

        if r:
            try:
                return int(s) == 1
            except:
                pass

        return False
